controllers/base_controller.go

```diff
@@ -98,47 +98,103 @@ func (r *BaseReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+		}
+		myEntryId = createEntryIfNotExistsResponse.Entry.EntryId
+	} else {
+		// One of the existing entries might already be just right for me... if not, we can choose one and update it
+		requiresUpdate := true
+		for _, entry := range matchedEntries {
+			if r.entryEquals(myEntry, entry) {
+				reqLogger.V(1).Info("Found existing identical enough spire entry", "entry", entry.EntryId)
+				myEntryId = entry.EntryId
+				requiresUpdate = false
+				break
+			}
+		}
+		if requiresUpdate {
+			reqLogger.V(1).Info("Updating existing spire entry to match desired state", "entry", matchedEntries[0].EntryId)
+			// It's important that if multiple instances are running they all pick the same entry here, otherwise
+			// we could have two instances of the registrar delete each others changes. This can only happen if both are
+			// also working off an up to date cache (otherwise the lagging one will pick up the other change later and correct
+			// the mistake.) If that happens then as long as the list of entries is always in the same order, we can guarantee they
+			// wont end up deleting all the entries and not noticing.
+			// TODO: Check that spire-server provides a stable sort order, or sort it ourselves!
+			myEntryId = matchedEntries[0].EntryId
+
+			myEntry.EntryId = myEntryId
+			_, err := r.SpireClient.UpdateEntry(ctx, &registration.UpdateEntryRequest{
+				Entry: myEntry,
+			})
+			if err != nil {
+				reqLogger.Error(err, "Failed to update existing spire entry", "existingEntry", matchedEntries[0].EntryId)
+				return ctrl.Result{}, err
+			}
+		}
 	}
 
-	err = r.deleteAllEntriesExcept(ctx, reqLogger, matchedEntries, createEntryIfNotExistsResponse.Entry.EntryId)
+	err = r.deleteAllEntriesExcept(ctx, reqLogger, matchedEntries, myEntryId)
 
 	return ctrl.Result{}, err
 }
 
-func (r *BaseReconciler) deleteAllEntries(ctx context.Context, reqLogger logr.Logger, entryIds []string) error {
-	for _, entry := range entryIds {
-		err := r.ensureDeleted(ctx, reqLogger, entry)
+func (r *BaseReconciler) makeEntryForObject(obj ObjectWithMetadata) *common.RegistrationEntry {
+	spiffeId := r.makeSpiffeId(obj)
+	parentId := r.makeParentId(obj)
+
+	if spiffeId == "" || parentId == "" {
+		return nil
+	}
+
+	return &common.RegistrationEntry{
+		Selectors: r.getSelectors(types.NamespacedName{
+			Namespace: obj.GetNamespace(),
+			Name:      obj.GetName(),
+		}),
+		ParentId: parentId,
+		SpiffeId: spiffeId,
+	}
+}
+
+func (r *BaseReconciler) entryEquals(myEntry *common.RegistrationEntry, b *common.RegistrationEntry) bool {
+	// TODO: Maybe this should be stricter on the other fields, but right now if you're editing entries on the server, I'll accept that odd things happen to you.
+	return b.SpiffeId == myEntry.GetSpiffeId()
+}
+
+func (r *BaseReconciler) deleteAllEntries(ctx context.Context, reqLogger logr.Logger, entries []*common.RegistrationEntry) error {
```

Comment:
Add comment that an error can leave the deletions in partially executed state?

Reply:
Rewritten to continue execution of deletions